src/test/operators/join_multi_predicates_test.cpp

```diff
@@ -0,0 +1,368 @@
+TYPED_TEST_CASE(JoinMultiPredicateTest, JoinMultiPredicateTypes, );  // NOLINT(whitespace/parens)
+
+TYPED_TEST(JoinMultiPredicateTest, InnerLTableSmallerRTableRandomNullsEqGt) {
+  this->_test_join_output(this->_base_choice_join_parameters.value());
+}
+
+TYPED_TEST(JoinMultiPredicateTest, LeftLTableSmallerRTableRandomNullsEqGt) {
+  auto parameters = this->_base_choice_join_parameters.value();
+  parameters.join_mode = JoinMode::Left;
+  parameters.expected_result_table_file_path =
+      "resources/test_data/tbl/join_operators/multi_predicates/"
+      "result_left_a_nulls_random_b_nulls_random_larger_eq_gt.tbl";
+  this->_test_join_output(parameters);
+}
+
+TYPED_TEST(JoinMultiPredicateTest, RightLTableSmallerRTableRandomNullsEqGt) {
+  auto parameters = this->_base_choice_join_parameters.value();
+  parameters.join_mode = JoinMode::Right;
+  parameters.expected_result_table_file_path =
+      "resources/test_data/tbl/join_operators/multi_predicates/"
+      "result_right_a_nulls_random_b_nulls_random_larger_eq_gt.tbl";
+  this->_test_join_output(parameters);
+}
+
+TYPED_TEST(JoinMultiPredicateTest, OuterLTableSmallerRTableRandomNullsEqGt) {
+  auto parameters = this->_base_choice_join_parameters.value();
+  parameters.join_mode = JoinMode::FullOuter;
+  parameters.expected_result_table_file_path =
+      "resources/test_data/tbl/join_operators/multi_predicates/"
+      "result_outer_a_nulls_random_b_nulls_random_larger_eq_gt.tbl";
+  // JoinHash does not support full outer joins
+  EXPECT_THROW(this->_test_join_output(parameters), std::logic_error);
+}
+
+TYPED_TEST(JoinMultiPredicateTest, RightLTableLargerRTableRandomNullsEqGt) {
+  auto parameters = this->_base_choice_join_parameters.value();
+  parameters.join_mode = JoinMode::Right;
+  parameters.table_pair.first = parameters.table_pair.second;
+  parameters.table_pair.second = this->_base_choice_join_parameters->table_pair.first;
+  // swap column pairs of the predicates
+  parameters.first_predicate.column_ids.first = this->_column_pair_1.second;
+  parameters.first_predicate.column_ids.second = this->_column_pair_1.first;
+  parameters.additional_predicates = {
+      {ColumnIDPair{this->_column_pair_2.second, this->_column_pair_2.first}, PredicateCondition::GreaterThan}};
+  parameters.expected_result_table_file_path =
+      "resources/test_data/tbl/join_operators/multi_predicates/"
+      "result_right_a_larger_nulls_rand_b_nulls_rand_eq_gt.tbl";
+  this->_test_join_output(parameters);
+}
+
+TYPED_TEST(JoinMultiPredicateTest, SemiLTableSmallerRTableRandomNullsEqGt) {
+  auto parameters = this->_base_choice_join_parameters.value();
+  parameters.join_mode = JoinMode::Semi;
+  parameters.expected_result_table_file_path =
+      "resources/test_data/tbl/join_operators/multi_predicates/"
+      "result_semi_a_nulls_random_b_nulls_random_larger_eq_gt.tbl";
+  this->_test_join_output(parameters);
+}
+
+TYPED_TEST(JoinMultiPredicateTest, AntiLTableSmallerRTableRandomNullsEqGt) {
```

Comment:
AntiRetainNulls and AntiDiscardNulls distinction still missing?

Reply:
no longer: 20a852a970cc9cd2cbbb204b89bcfbbd62fc9de9